Improved documentation for join protocol (#2497)
microsoft/CCF: doc/operations/start_network.rst

The joining node takes the certificate of the existing network to join via ``--network-cert-file`` and initiates an enclave-to-enclave TLS connection to an existing node of the network as specified by ``--target-rpc-address``.

A new node can only join an existing CCF network if its SGX quote is valid  [#remote_attestation]_. and runs an enclave application that is :ref:`trusted by the consortium <governance/common_member_operations:Updating Code Version>`.

If the network has not yet been opened by members (see :ref:`governance/open_network:Opening the Network`), the joining node becomes part of the network immediately.

Otherwise, if the network has already been opened to users, members need to trust the joining node before it can become part of the network and participate in the consensus (see :ref:`governance/common_member_operations:Trusting a New Node`).

The ``Pending`` joining node will automatically poll the service (interval configurable via ``--join-timer`` option) until the members have successfully transitioned the node to the ``Trusted`` state. It is only then that the joining node will transition to the ``PartOfNetwork`` state and start updating its ledger.

> .. tip:: After the node has been trusted by members, operators should poll the ``/node/state`` endpoint on the newly added node until the ``{"state": "PartOfNetwork"}`` is reported. This status confirms that the replication of the ledger has started on this node.

Reviewer: Should we talk about the difference between `PartOfNetwork` and "fully caught up" - ie, you may want to poll the node's commit index and only direct users to it once it has _finished_ replication?

Reviewer 2: What does "finished" mean?

Author: Yeah, it's awkward to define "finished" here. If we decide to go the "[learner](https://etcd.io/docs/v3.3/learning/learner/#features-in-v35)" way at some point, we could have some thresholds to define when a node actually becomes part of the service (e.g. `N` transactions after it's seen its own addition to the store).

Reviewer 2: I would prefer it not be mentioned until we have a proper definition for it.

Reviewer: I think finished is up to the operator, but if we're clarifying the meaning of these states we should also clarify that a `PartOfNetwork` node may still be serving reads at `1.10`, despite your service being at `10.999999`?

Reviewer 2: Sure, but that's always true, nodes can lag behind etc.

Reviewer: Sure, so if you were monitoring all of your nodes and noticed it was lagging, you may want to remove it from your load balancer while you investigated. This lag is going to be most pronounced, and always present even in a well-behaved network, on a new joiner.

Reviewer 2: If they join from a recent snapshot, it's going to be quite slight and rapidly plugged. Adding a reminder is fine, and the operator may have a QoS policy, but clients still need to be robust to followers lagging.

Virtual Mode
------------

To start a CCF node in `virtual` mode, operators should use the `virtual` variants of the ``cchost`` binary and enclave application:

.. code-block:: bash

    $ cchost.virtual --enclave-file /path/to/virtual_enclave_library [args]

.. warning:: Nodes started in virtual mode provide no security guarantees. They should never be used for production purposes.

Signature Interval
------------------

Transaction commit latency in a CCF network is primarily a function of signature frequency. A network emitting signatures more frequently will be able to commit transactions faster, but will spend a larger proportion of its execution resources creating and verifying signatures. Setting signature frequency is a trade-off between transaction latency and throughput.

Two options are provided to that end:

- ``--sig-tx-interval``: number of transactions between two signatures
- ``--sig-ms-interval``: time in milliseconds between two signatures

> .. note:: These options specify the intervals at which the generation of signature transactions is `triggered`. However, because of the parallel execution of transactions, it is possible that signature transactions are recorded in the ledger at a slightly higher interval than the specified values.

Reviewer: ```suggestion
.. note:: These options specify the intervals at which the generation of signature transactions is `triggered`. However, because of the parallel execution of transactions, the actual intervals between signature transactions may be slightly larger.
```

I think "at a slightly higher interval" is odd wording, prefer _larger_ interval. Reworded to clarify this.
